oculus_ros2/include/oculus_ros2/sonar_viewer.hpp

```diff
@@ -0,0 +1,138 @@
+  // sensor_msgs::msg::Image publish_fan(const oculus::PingMessage::ConstPtr &ping);
+
+private:
+  //      //   sensor_msgs::ImagePtr msg;
+  const rclcpp::Node* node_;
+  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_publisher_;
+};
+
+template <typename T>
+std::vector<double> linspace(T start_in, T end_in, int num_in) {
+  std::vector<double> linspaced;
+
+  double start = static_cast<double>(start_in);
+  double end = static_cast<double>(end_in);
+  double num = static_cast<double>(num_in);
+
+  if (num == 0) {
+    return linspaced;
+  }
+  if (num == 1) {
+    linspaced.push_back(start);
+    return linspaced;
+  }
+
+  double delta = (end - start) / (num - 1);
+
+  for (int i = 0; i < num - 1; ++i) {
+    linspaced.push_back(start + delta * i);
+  }
+  linspaced.push_back(end);  // I want to ensure that start and end
+                             // are exactly the same as the input
+  return linspaced;
+}
+
+template <typename dataType>
+void SonarViewer::publish_fan(const int& width,
+    const int& height,
+    const int& offset,
+    const std::vector<dataType>& ping_data,
+    const int& master_mode,
+    const double& ping_range,
+    const std_msgs::msg::Header& header) const {
+  // Create rawDataMat from ping_data
+  cv::Mat rawDataMat(height, (width + 4), CV_8U);
+  std::memcpy(rawDataMat.data, ping_data.data() + offset, height * (width + 4));
+
+  int bearing = (master_mode == 1) ? 65 : 40;
+  std::vector<double> ranges = linspace(0., ping_range, height);
+  int image_width = 2 * std::sin(bearing * M_PI / 180) * ranges.size();
+  cv::Mat mono_img;
+  if (std::is_same<dataType, uint8_t>::value) {
+    mono_img = cv::Mat::ones(cv::Size(image_width, ranges.size()), CV_8UC1) * 255;
+  } else {
+    mono_img = cv::Mat::ones(cv::Size(image_width, ranges.size()), CV_16UC1) * 65535;
+  }
+
+  // for (int i = 0; i < image_width; i++) {
+  //   for (int j = 0; j < ranges.size(); j++) {
+  //     mono_img.at<dataType>(j, i) = 255;
+  //   }
+  // }
+
+  const float ThetaShift = 1.5 * 180;
+  const cv::Point origin(image_width / 2, ranges.size());
+
+  for (int r = 0; r < ranges.size(); r++) {  // TODO(??, optimize for cuda)
+    std::vector<cv::Point> pts;
+    cv::ellipse2Poly(origin, cv::Size(r, r), ThetaShift, -bearing, bearing, 1, pts);
+
+    std::vector<cv::Point> arc_points;
+    arc_points.push_back(pts[0]);
+
+    for (size_t k = 0; k < (pts.size() - 1); k++) {  // TODO(??, optimize for cuda)
+      cv::LineIterator it(mono_img, pts[k], pts[k + 1], 4);
+      for (int i = 1; i < it.count; i++, ++it) arc_points.push_back(it.pos());  // TODO(??, optimize for cuda)
+    }
+
+    cv::Mat data_rows_resized;
+    cv::resize(rawDataMat.row(r), data_rows_resized, cv::Size(arc_points.size(), arc_points.size()));
+
+    for (size_t k = 0; k < arc_points.size(); k++)  // TODO(??, optimize for cuda)
+      mono_img.at<dataType>(arc_points[k]) = data_rows_resized.at<dataType>(1, k);
+  }
```

Comment:
I profiled the code and as expected, this loop is the most consuming. Without adding extra dependency, you can parallelize the code on CPU. Use `cv::parallel_for_`:

```
  cv::parallel_for_(cv::Range(0, ranges.size()), [&](const cv::Range& range) {
    for (int r = range.start; r < range.end; r++) {
      //double r = ranges[i];
      std::vector<cv::Point> pts;
      cv::ellipse2Poly(origin, cv::Size(r, r), ThetaShift, -bearing, bearing, 1, pts);
      
			std::vector<cv::Point> arc_points;
			arc_points.push_back(pts[0]);

			for (size_t k = 0; k < (pts.size() - 1); k++) {  // TODO(??, optimize for cuda)
				cv::LineIterator it(mono_img, pts[k], pts[k + 1], 4);
				for (int i = 1; i < it.count; i++, ++it) arc_points.push_back(it.pos());  // TODO(??, optimize for cuda)
			}

			cv::Mat data_rows_resized;
			cv::resize(rawDataMat.row(r), data_rows_resized, cv::Size(arc_points.size(), arc_points.size()));

			for (size_t k = 0; k < arc_points.size(); k++)  // TODO(??, optimize for cuda)
				mono_img.at<dataType>(arc_points[k]) = data_rows_resized.at<dataType>(1, k);            
    }
    });
```